server/src/state.rs

```diff
@@ -168,6 +169,25 @@ impl NodesState {
         Ok(())
     }
 
+    pub fn check_request_duplicate(
+        &self,
+        id: SessionId,
+        request_id: RequestId,
+    ) -> ServerResult<bool> {
+        match self.sessions.get(&id) {
+            None => Err(Unauthorized::SessionNotFound(id).into()),
+            Some(&slot) => match self.slots.get(slot as usize) {
+                Some(Slot::Some(node)) => {
+                    let result = node.request_history.contains(request_id);
+                    node.request_history.push(request_id);
```

Comment:
Are you sure that this push actually modifies `self.sessions.slots.node`?

Reply:
Yes, I've tested the change. The internal `VecDeque` is behind `Arc<Mutex<_>>`.